src/express.re

```diff
@@ -1,43 +1,149 @@
-/*
-  There are a few types to update in this file.
-  We're generic in 'a in most types, but really
-  should specify which JS type should be used.
-
-  An example of how to handle JS fns that take
-  more than one type, where some types are
-  containers and some primitives is the the
-  Response.send binding below.
-
-  sendString | sendJson | sendBuffer | sendArray
-  are all bindings to the .send method based on the
-  input type
-
-  tracked in https://github.com/BuckleTypes/bs-express/issues/6
- */
-module Next = {
+type done_;
+/** abstract type which ensure middleware function must either
+    call the [next] function or one of the [send] function on the 
+    response object. 
+
+    This should be a great argument for OCaml, the type system 
+    prevents silly error which in this case would make the server hang */ 
+
+/** TODO : maybe this should be a more common module like bs-error */
+module Error = {
+
   type t;
+  /** Error type */
+
+  external message : t => option string = "" 
+    [@@bs.send][@@bs.return null_undefined_to_opt];
+  external name : t => option string = "" 
+    [@@bs.send][@@bs.return null_undefined_to_opt];
 };
 
 module Request = {
   type t;
+  
+  type params = Js_dict.t Js_json.t;
+
+  external params : t => params = "" [@@bs.get];
+  /** [params request] return the JSON object filled with the 
+      request parameters */
+
+  external asJSONObject : t => Js_dict.t Js_json.t = "%identity";
```

Comment:
This name is inconsistent with `Response.sendJson`. I don't have a strong opinion on which is better, but it should be one or the other.